### DISCUSSION TITLE
Versioning - Pre-Release Banner

### DISCUSSION THREAD

**Discussion by adam-moss** [squidfunk/mkdocs-material#5803] [Q&A] [answered]
Hi,

New to mkdocs and mkdocs-material, both of which I'm finding most awesome with an amazingly simple learning curve, and have inspired me to create an additional handler for `mkdocstrings`.

Apologies if this has been asked/answered but my search foo is failing me.

In https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/#versioning it describes how to implement a customisation to incorporate an "outdated" banner in my docs.  This works an absolute treat, although I did have to spend a bit of trial-and-error with `mike` as I'm using GitLab rather than GitHub (and a bit OCD).

Now I'm wondering if it is possible to basically do the opposite of the outdated banner, and highlight when the user is on a "future" / pre-release version of the docs?

At the moment I publish updated docs as `next` when merged to `main`, and publish a tagged version with the `latest` alias only when a release is tagged and packaged.

Thusly it would be nice to highlight these docs as being pre-release.

Thanks.

**Comment by squidfunk** [comment]
Thanks for asking, and we're glad that you enjoy our work! There's currently no pre-release banner, and it's also not planned right now, but you could just use the [announcement bar](https://squidfunk.github.io/mkdocs-material/setup/setting-up-the-header/#announcement-bar) with some custom styling and only enable it on pre-release builds? How? That's left to be worked out, but maybe ask the authors of mike whether you can detect if you're running a pre-release build in the template. You could then define a custom announcement bar and only render it when you're on a pre-release build. I understand that this sounds involved, and I'm sorry that there's currently no easy way to do this, but it should be possible with some legwork.

We will consider this feature in the future.

**Reply by adam-moss** [reply]
Ah... that's a great idea, thank you, I'll have a play around share what I come up with 😄

**Reply by adam-moss** [reply, marked as answer]
I've just had a chance to play with this and have come up with a fairly clean solution I think.

In `mkdocs.yml`, add an `extra` attribute that is passed into the `jinja2` template via `mkdocs` itself:

```yaml
extra:
  pre_release: true
```

In `overrides/main.html`, use the passed variable in a conditional surrounding the `announce` block:

```jinja2
{% extends "base.html" %}

{% if config.extra.pre_release %}
  {% block announce %}
    You are currently viewing <strong>pre-release</strong> documentation.  This may contain features that have not yet been packaged. For the most accurate information, please <strong><a href="{{ '../' ~ base_url }}">access the release version</a></strong>.
  {% endblock %}
{% endif %}

{% block outdated %}
  You are currently viewing <strong>outdated</strong> documentation.  For the most up-to-date information, please <strong><a href="{{ '../' ~ base_url }}">refer to the latest release</a></strong>.
{% endblock %}
```

And in `extra.css`, change `.md-banner` to match the `md-banner--warning` styling:

```css
.md-banner {
  background-color: var(--md-warning-bg-color);
  color: var(--md-warning-fg-color);
}
```

I must admit I'm not entirely sure on the implications of that last one elsewhere in the theme, looks through the `src` folder doesn't suggest anything obvious but it is easy enough to remove if needs be.

Thanks for the suggestion pointing me in the right direction 😄

**Reply by squidfunk** [reply]
To be honest, I wasn't sure if conditional definition of blocks works, so cool! 

> I must admit I'm not entirely sure on the implications of that last one elsewhere in the theme, looks through the src folder doesn't suggest anything obvious but it is easy enough to remove if needs be.

You're using the announcement bar feature as advertised, with the slight change that you can toggle it via `mkdocs.yml`, so all should be good. To me, this looks like a very clean solution, also the CSS 🚀

**Reply by AdrianoKF** [reply]
I just tried this out, however, it seems that conditionally defining blocks in Jinja isn't supported - I could not get the notification to be hidden, despite a lot of attempts.

What (kinda) worked for me was flipping the template around: defining the `announce` block unconditionally, and guarding the rendering of its content with an `if`. However, when the toggle is turned off, the base template then unfortunately renders the banner containing an inner div with a margin (which shows up above the main header as a bit of empty space). With a few CSS hacks I got to a point where toggling the banner yields the desired outcome, but it's not pretty.

```jinja
{% extends "base.html" %}
{% block announce %}
  {% if config.extra.pre_release %}
    <!-- Need to reapply margin from base CSS, which is overridden in extra CSS (to fix empty banner) -->
    <div style="margin: 0.6rem auto">
      ...
    </div>
  {% endif %}
{% endblock %}
```

```css
.md-banner__inner {
    margin: 0;
}
```

```yaml
extra:
  pre_release: !ENV [DOCS_PRERELEASE, false]
```

**Reply by squidfunk** [reply]
If you cleverly use Jinja whitespace control, the banner should not be rendered, as the `announce` block will be empty:

``` jinja
{% extends "base.html" %}
{% block announce %}
  {%- if config.extra.pre_release -%}  <!-- See the dashes here -->
    <!-- Need to reapply margin from base CSS, which is overridden in extra CSS (to fix empty banner) -->
    <div style="margin: 0.6rem auto">
      ...
    </div>
  {%- endif -%} <!-- See the dashes here -->
{% endblock %}
```

**Reply by AdrianoKF** [reply]
Thanks, @squidfunk, that does the trick (TIL about the whitespace control in Jinja - I was familiar with the idea from Helm charts).

In this case, all the stylesheets in both the template and the extra CSS file can be omitted 🙌🏻

**Reply by adam-moss** [reply]
After a lot of testing with it inconjunction with `mike`, on GitLab & GitLab Pages, I ended up with 

```plaintext
theme
 |- main.html
 |- main.jinja
 |- outdated-warning.jinja
 |- pre-release-warning.jinja
```

with content:

`main.html`
```html
{% extends "main.jinja" %}
```

`main.jinja`
```jinja
{% if config.extra.show_pre_release_warning %}
  {% include "pre-release-warning.jinja" %}
{% else %}
  {% include "outdated-warning.jinja" %}
{% endif %}
````

`outdated-warning.jinja`
```jinja
{% extends "base.html" %}

{% block outdated %}
  You are currently viewing <strong>outdated</strong> documentation.  For the most up-to-date information, please <strong><a href="{{ '../' ~ base_url }}">refer to the latest release</a></strong>.
{% endblock %}
```

`pre-release-warning.jinja`
```jinja
{% extends "base.html" %}

{% block announce %}
  You are currently viewing <strong>pre-release</strong> documentation.  This may contain features that have not yet been packaged. For the most accurate information, please <strong><a href="{{ '../' ~ base_url }}">refer to the latest release</a></strong>.
{% endblock %}
```

Mostly because I couldn't get the whitespace to play nice in all scenarios.

**Reply by staadecker** [reply]
@adam-moss How did you manage to prevent the outdated banner from displaying on your `next` site. My `mkdocs.yml` is as follows but the banner still displays (so I get both the outdated and the pre-release banners).

```yml
extra:
  pre_release: !ENV [DOCS_PRERELEASE, true]
  version:
    #  See: https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/
    provider: mike
    default:
      - latest
      - next # included so that there's no outdated warning, but not working
    alias: true
```

Fyi, it would be nice to at least have a link in the documentation pointing to this thread @squidfunk !

**Reply by squidfunk** [reply]
> Fyi, it would be nice to at least have a link in the documentation pointing to this thread @squidfunk !

Could you please clarify what you mean with _at least_? This is not a feature that we committed to support. It's a cutomization through and through. Yes, we consider making it simpler to add custom banners in the future, but right now, you're basically on your own when customizing. I hope you can understand, that we as maintainers have to balance what we support natively and what we leave for customization by users. Thus, I think a link in the documentation would endorse that we built this, which we didn't. It's a customization created by @adam-moss.

**Reply by staadecker** [reply]
@squidfunk my apologies if the *at least* came off as presumptuous! I fully understand why you might not want to natively support this feature. I guess I was just trying to say that if you're ever considering adding native support, you have my +1.

Regarding the message, perhaps something like [this](https://github.com/squidfunk/mkdocs-material/pull/8455) strikes a good balance. I think having a link (while making it clear that there is no endorsement) could save people (like me) a lot of time! Your call, it's just a suggestion.

**Reply by staadecker** [reply]
> My `mkdocs.yml` is as follows but the banner still displays (so I get both the outdated and the pre-release banners).

Turns out I needed to clear the `/next/.__outdated` session storage after updating my `mkdocs.yml` to get the webpage to recalculate whether `next` is outdated. All is good now!
